frame/screencopy_tracker.py

```diff
@@ -0,0 +1,133 @@
+        elif iface_name == WlOutput.name:
+            self.output = registry.bind(id_num, WlOutput, version)
+        elif iface_name == WlShm.name:
+            self.shm = shm = registry.bind(id_num, WlShm, version)
+
+    def connected(self) -> None:
+        self.copy_frame(True)
+
+    def disconnected(self) -> None:
+        if self.shm_data is not None:
+            self.shm_data.close()
+            self.shm_data = None
+
+    def _frame_buffer(self, frame, format: int, width: int, height: int, stride: int) -> None:
+        assert self.buffer_width == 0 or self.buffer_width == width, 'Buffer width changed'
+        self.buffer_width = width
+        assert self.buffer_height == 0 or self.buffer_height == height, 'Buffer height changed'
+        self.buffer_height = height
+        buffer_size = stride * height
+        assert self.buffer_size == 0 or self.buffer_size == buffer_size, 'Buffer size changed'
+        self.buffer_size = buffer_size
+        if self.buffer is None:
+            assert self.shm is not None, 'SHM not created'
+            fd = shm_open()
+            os.ftruncate(fd, buffer_size)
+            self.shm_data = mmap.mmap(fd, buffer_size)
+            shm_pool = self.shm.create_pool(fd, buffer_size)
+            libc.close(fd);
+            self.buffer = shm_pool.create_buffer(0, width, height, stride, format)
+            shm_pool.destroy()
+
+    def _frame_damage(self, frame, x: int, y: int, width: int, height: int) -> None:
+        self.pending_damage += width * height
+
+    def _frame_ready(self, frame, tv_sec_hi, tv_sec_lo, tv_nsec) -> None:
+        self.frame_count += 1
+        self.total_damage += (self.pending_damage
+            if self.pending_damage
+            else (self.buffer_width * self.buffer_height))
+        self.pending_damage = 0
+        assert self.frame is not None, 'Frame is None'
+        assert self.display is not None, 'No display'
+        self.copy_frame(False)
+        self.display.flush()
+
+    def copy_frame(self, is_initial: bool) -> None:
+        assert self.screencopy_manager is not None, f'{ZwlrScreencopyManagerV1.name} not supported'
+        assert self.display is not None, 'No display'
+        if self.frame is not None:
+            self.frame.destroy()
+        self.frame = frame = self.screencopy_manager.capture_output(0, self.output)
+        frame.dispatcher['buffer'] = self._frame_buffer
+        frame.dispatcher['damage'] = self._frame_damage
+        frame.dispatcher['ready'] = self._frame_ready
+        self.display.roundtrip()
+        assert self.buffer is not None, 'No buffer info given'
+        if is_initial:
+            frame.copy(self.buffer)
+        else:
+            frame.copy_with_damage(self.buffer)
+
+    def properties(self) -> dict[str, Any]:
+        total_possible_pixels = max(
+            self.frame_count * self.buffer_width * self.buffer_height,
+            self.total_damage,
+            1
```

Comment:
Why 1?

Reply:
So we don't end up dividing by zero and crashing if nothing gets set

Reply:
D'uh, should've looked further down ;)